### PR TITLE
Fix host files ownership

### PR DESCRIPTION
**Pull request by thir820**
# Changes

- Fix and improve the ownership and mode of copied host files.
- Start documentation for root generator.

# Dependencies:

none

# Tests results

Unit tests:
```bash
ebcl@4fcab0ed6b55:/workspace$ pytest
================================================================================================= test session starts ==================================================================================================
platform linux -- Python 3.10.12, pytest-8.3.3, pluggy-1.5.0
rootdir: /workspace
configfile: pyproject.toml
testpaths: tests
plugins: libtmux-0.37.0, cov-5.0.0
collected 74 items                                                                                                                                                                                                     

tests/test_apt.py .......                                                                                                                                                                                        [  9%]
tests/test_boot.py ..                                                                                                                                                                                            [ 12%]
tests/test_cache.py ....                                                                                                                                                                                         [ 17%]
tests/test_config.py .....                                                                                                                                                                                       [ 24%]
tests/test_deb.py ...                                                                                                                                                                                            [ 28%]
tests/test_dependency.py ..                                                                                                                                                                                      [ 31%]
tests/test_fake.py ....                                                                                                                                                                                          [ 36%]
tests/test_files.py ...................                                                                                                                                                                          [ 62%]
tests/test_initrd.py ........                                                                                                                                                                                    [ 72%]
tests/test_proxy.py .......                                                                                                                                                                                      [ 82%]
tests/test_root.py ..s...s.                                                                                                                                                                                      [ 93%]
tests/test_version.py .....                                                                                                                                                                                      [100%]

====================================================================================== 72 passed, 2 skipped in 1039.75s (0:17:19) ======================================================================================
```

Robot tests:

```bash
 ebcl@4fcab0ed6b55:/workspace/robot_tests$ robot *.robot
==============================================================================
Boot & Initrd & Root                                                          
==============================================================================
Boot & Initrd & Root.Boot                                                     
==============================================================================
Kernel exists                                                         | PASS |
------------------------------------------------------------------------------
Config is OK                                                          | PASS |
------------------------------------------------------------------------------
Script was executed                                                   | PASS |
------------------------------------------------------------------------------
Boot & Initrd & Root.Boot                                             | PASS |
3 tests, 3 passed, 0 failed
==============================================================================
Boot & Initrd & Root.Initrd                                                   
==============================================================================
Root Device is Mounted                                                | PASS |
------------------------------------------------------------------------------
Devices are Created                                                   | PASS |
------------------------------------------------------------------------------
Rootfs should be set up                                               | PASS |
------------------------------------------------------------------------------
File dummy.txt should be OK                                           | PASS |
------------------------------------------------------------------------------
File other.txt should be OK                                           | PASS |
------------------------------------------------------------------------------
Boot & Initrd & Root.Initrd                                           | PASS |
5 tests, 5 passed, 0 failed
==============================================================================
Boot & Initrd & Root.Root                                                     
==============================================================================
Systemd should exist                                                  | PASS |
------------------------------------------------------------------------------
Fakeoot config executed                                               | PASS |
------------------------------------------------------------------------------
Fakechroot config was executed                                        | PASS |
------------------------------------------------------------------------------
Chroot config was executed                                            | PASS |
------------------------------------------------------------------------------
Boot & Initrd & Root.Root                                             | PASS |
4 tests, 4 passed, 0 failed
==============================================================================
Boot & Initrd & Root                                                  | PASS |
12 tests, 12 passed, 0 failed
==============================================================================
Output:  /workspace/robot_tests/output.xml
Log:     /workspace/robot_tests/log.html
Report:  /workspace/robot_tests/report.html
```

# Developer Checklist:

- [x] Test: Changes are tested
- [x] Doc: Documentation has been updated 
- [x] Git: Informative git commit message(s)
- [x] Issue: If a related GitHub issue exists, linking is done

## Checklists for documentation

- [x] Grammar: the content is grammatically correct
      (spelling, grammar, formatting, US English is used)
- [x] Comprehensibility/Unambiguousness: the content is easy readable, clear to understand
- [x] Correctness and consistency: the content is technically correct and consistent,
      no contradictions, no double descriptions
- [x] Terminology: technical terms are clear and they are used correctly and documented in the glossary
- [x] Level of detail: the content matches the detail level necessary for the reviewed object

# Reviewer checklist:
- [x] Review: Changes are reviewed
- [ ] Review: Tested by the reviewer

## Checklists for documentation
- [x] Grammar: the content is grammatically correct
      (spelling, grammar, formatting, US English is used)
- [x] Comprehensibility/Unambiguousness: the content is easy readable, clear to understand
- [x] Correctness and consistency: the content is technically correct and consistent,
      no contradictions, no double descriptions
- [x] Terminology: technical terms are clear and they are used correctly and documented in the glossary
- [x] Level of detail: the content matches the detail level necessary for the reviewed object
0 commit comments
Comments
0

### CODE REVIEW

**Review by simone-weiss — DISMISSED** [review]

File: ebcl/common/files.py [hunk 1]
```diff
@@ -178,14 +178,14 @@ def copy_file(
...
+                    # Change owner to host user and group.
                     self.fake.run_sudo(
                         f'chown -R {os.getuid()}:{os.getgid()} {file}')
-                    self.fake.run_sudo(f'chmod -R 755 {file}')
 
-                if os.path.isfile(file):
-                    self._run_cmd(
-                        f'mkdir -p {os.path.dirname(target)}',
-                        environment, check=False)
+                # Create target directory if it not exists.
```
Comment (MofX) [inline ebcl/common/files.py:185]:
```suggestion
                # Create target directory if it does not exist
```

File: ebcl/common/files.py [hunk 2]
```diff
@@ -210,6 +211,12 @@ def copy_file(
                     self._run_cmd(f'chown {uid} {target}', environment)
                 if gid:
                     self._run_cmd(f'chown :{gid} {target}', environment)
+
+                if not mode and not move:
+                    # Take over mode from source file.
+                    mode = oct(os.stat(file).st_mode)
+                    mode = mode[-3:]
```
Comment (MofX) [inline ebcl/common/files.py:218]:
```suggestion
                    mode = mode[-4:]
```
This keeps the suid, sgid and sticky bit as well

File: tests/test_root.py [hunk 3]
```diff
@@ -55,6 +55,7 @@ def test_build_kiwi_image(self):
         assert archive
         assert os.path.isfile(archive)
 
+    @pytest.mark.skip(reason="Elbe not part of dev container anymore.")
```
Comment (MofX) [inline tests/test_root.py:58]:
```suggestion
    @pytest.mark.skip(reason="Elbe is not part of dev container anymore.")
```

File: tests/test_root.py [hunk 4]
```diff
@@ -107,6 +108,7 @@ def test_build_sysroot_kiwi(self):
...
```
Comment (MofX) [inline tests/test_root.py:111]:
```suggestion
    @pytest.mark.skip(reason="Elbe is not part of dev container anymore.")
```

**Review by MofX — CHANGES_REQUESTED** [review]
And typo in commit message: "fix file ownership hanlding"

**Comment by thir820** [comment]
> And typo in commit message: "fix file ownership hanlding"

Yes, saw it too late.

**Comment by MofX** [comment]
> > And typo in commit message: "fix file ownership hanlding"
> 
> Yes, saw it too late.

Just fix it using force-pushed. IMHO this PR shouldn't be squashed anyway (or squash merges in generally should never be used), because it renders the git history completely useless especially for unconnected commits like in this PR

**Comment by thir820** [comment]
> > > And typo in commit message: "fix file ownership hanlding"
> > 
> > 
> > Yes, saw it too late.
> 
> Just fix it using force-push. IMHO this PR shouldn't be squashed anyway (or squash merges in generally should never be used), because it renders the git history completely useless especially for unconnected commits like in this PR

I don't get this. No I have several commits for my changes and all the review findings, and the squash and merge will make one commit on main out of it. This makes sense for me, I and don't know why the separate commits of this PR should be preserved.

**Comment by MofX** [comment]
> I don't get this. No I have several commits for my changes and all the review findings, and the squash and merge will make one commit on main out of it. This makes sense for me, I and don't know why the separate commits of this PR should be preserved.

A commit should be self-contained and do one job. When you squash this pr, the commit does not fulfil the second requirement anymore

And a commit message should describe what was changed. This is how every serious open source project uses git.
I don't want to preserve all the useless fixup commits, I suggest to always make changes, force push clean commits and then merge-commit or forward-merge the changes in the end. This way the commit history has value. With every other workflow the commit history and commit messages have no value at all (take a look at this useless commit message as an example:https://github.com/Elektrobit/ebcl_build_tools/commit/9483769aa215f0cc250a1fef16f846976b91dd0a)

**Comment by thir820** [comment]
> > I don't get this. No I have several commits for my changes and all the review findings, and the squash and merge will make one commit on main out of it. This makes sense for me, I and don't know why the separate commits of this PR should be preserved.
> 
> A commit should be self-contained and do one job. When you squash this pr, the commit does not fulfil the second requirement anymore
> 
> And a commit message should describe why something was changed. This is how every serious open source project uses git. I don't want to preserve all the useless fixup commits, I suggest to always make changes, force push clean commits and then merge-commit or forward-merge the changes in the end. This way the commit history has value. With every other workflow the commit history and commit messages have no value at all (take a look at this useless commit message as an example:[9483769](https://github.com/Elektrobit/ebcl_build_tools/commit/9483769aa215f0cc250a1fef16f846976b91dd0a))

It was done differently by now, but in general I agree. I did the squashing manually and force pushed.

**Pull request merged**